pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -1434,7 +1434,7 @@ protected void internalDeleteSubscription(AsyncResponse asyncResponse, String su
             internalDeleteSubscriptionForNonPartitionedTopic(asyncResponse, subName, authoritative);
         } else {
             getPartitionedTopicMetadataAsync(topicName,
-                    authoritative, false).thenAccept(partitionMetadata -> {
+                    authoritative, false).thenAcceptAsync(partitionMetadata -> {
```

Comment:
I think it's better to make `internalDeleteSubscriptionForNonPartitionedTopic` return CompletableFuture?

Reply:
I'm a bit confused, we're blocked here: https://github.com/apache/pulsar/blob/94736a43f1b9a6d1db75936032b94eb9a11b9c0d/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java#L1500
which is within `internalDeleteSubscriptionForNonPartitionedTopic` method, how will making change in the else block solve this issue?

Reply:
@Jason918 I don't understand how making `internalDeleteSubscriptionForNonPartitionedTopic` return CompletableFuture helps with solving the deadlock issue? The real issue here is the blocking `thenAccept` call in the metadata store callback thread. The lambda for `thenAccept` will get executed in the callback thread, but that does not happen due to the blocking `thenAccept`.

Reply:
By making it return CompletableFuture, I mean to implement the async non-blocking version of `internalDeleteSubscriptionForNonPartitionedTopic`.

The root blocking operation in your case happens in `internalDeleteSubscriptionForNonPartitionedTopic`, like `validateTopicOwnership` , `validateTopicOperation` and even `getTopicReference`.

For sure, change it to `thenAcceptAsync` should solve your case. But if your topic is partitioned, the code will goes to Line 1434 above, the deadlock still will happen, right?

Reply:
It is OK to block in the call to `internalDeleteSubscriptionForNonPartitionedTopic` in Line 1434 as that code does not execute in the metadata-store callback thread (it executes in the REST handler).

In order to make `internalDeleteSubscriptionForNonPartitionedTopic` async non-blocking and have it return a completableFuture, the code changes are extensive. One has to make sure all the code in `internalDeleteSubscriptionForNonPartitionedTopic` non-blocking. That is `validateTopicOwnership`, `validateTopicOperation`, `sub.delete().get()` and all the methods that they in turn invoke. For e.g.:

https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java#L605

The scope is fairly large impacting a large swath of the code.

 It also doesn't solve the problem entirely, as there are other blocking calls in the `thenAccept` lambda. For e.g.,:

https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java#L1455

In the future, if some one adds a blocking call in the `thenAccept` lambda the problem will get re-introduced, which is very easily missed in a code review.

For the above reasons, changing `thenAccept` to `thenAcceptAsync` seems to be the best option for now.

Reply:
Thx for the detailed explanation.







> It is OK to block in the call to internalDeleteSubscriptionForNonPartitionedTopic in Line 1434 as that code does not execute in the metadata-store callback thread (it executes in the REST handler).

Agreed on "not execute in the metadata-store callback thread", but it's always not good to execute blocking calls on executor. Maybe it's a little out of the scope of this issue. 

> The scope is fairly large impacting a large swath of the code.

Actually, I think we already have most of them in async mode.  


> It also doesn't solve the problem entirely, as there are other blocking calls in the thenAccept lambda. For e.g.,:

Do you mean `waitForAll` ?  It's not blocking.

Reply:
At least, maybe we should use "pulsar().getExecutor()" to execute this `thenAcceptAsync`?

Reply:
I have used `pulsar().getExecutor()` per your suggestion. PTAL.